chain/chain/src/chain.rs

```diff
@@ -1797,7 +1804,7 @@ impl Chain {
         let (apply_chunk_work, block_preprocess_info) = preprocess_res;
 
         // 2) Start creating snapshot if needed.
-        if let Err(err) = self.process_snapshot() {
+        if let Err(err) = self.process_snapshot(block.header()) {
             tracing::error!(target: "state_snapshot", ?err, "Failed to make a state snapshot");
         }
```

Comment:
This is a v good question, I am not sure… I think it should be pretty possible to just decouple this, since it’s not needed in any block/chunk processing. My guess is it was just easy to express it as a callback that gets called every time we update the HEAD, in a way that you could sort of imagine the indexer being rewritten to do (instead of polling every so often for a new block)

Reply:
I'm fine having this check to happen on every block, it's a common pattern. It just seems strange that it happens before the block is applied. If the block turns out to be invalid the node may do an "invalid" snapshot for some definition of invalid.

Reply:
wait yea that is a good point.. and I feel like not only should we wait until it's applied, but maybe we should wait until it's final too

Reply:
I know we discussed it but I don't remember the final conclusion - can you remind me please?

Reply:
Oh I think it's just a little bit tricky where we have to make sure flat storage is still available when we make the snapshot, and that the flat storage head hasn't moved past the point where we want to take the snapshot. I'm not 100% sure what the guarantees are on that, but keeping this here at least preserves the old behavior, and in another PR if it makes sense we can move this. But worth noting that in one of the recent commits addressing your comment about not passing the block header to `Chain::should_make_or_delete_snapshot()`, we now take a snapshot for the current epoch sync hash one block later than before, and it still works